internal/dcache/rpc/response_error.go

```diff
@@ -72,26 +72,76 @@ func IsRPCError(err error) bool {
-func IsConnectionTerminated(err error) bool {
+//
+// Note: This can only be received when we are sending the RPC, i.e. only write()/send() can fail with this.
+//
+//	See IsConnectionReset() for similar error that read/recv can fail with.
+func IsBrokenPipe(err error) bool {
+	common.Assert(err != nil)
+
+	// RPC error, cannot be a broken pipe error.
+	if GetRPCResponseError(err) != nil {
+		log.Debug("IsBrokenPipe: is RPC error: %v", err)
+		return false
+	}
+
+	te := thrift.NewTTransportExceptionFromError(err)
+	// Note: This doesn't work.
+	//return te.TypeId() == thrift.NOT_OPEN
+
+	log.Debug("IsBrokenPipe: err: %v, err: %T, te.TypeId(): %d, Is syscall.EPIPE: %v",
+		err, err, te.TypeId(), errors.Is(err, syscall.EPIPE))
+
+	//
+	// In this case I've seen syscall.EPIPE check to work, but since it's not documented
+	// we do the string match also, just in case.
+	//
+	brokenPipe := "roken pipe"
+	return errors.Is(err, syscall.EPIPE) || strings.Contains(err.Error(), brokenPipe)
+}
+
+// This is the standard "connection reset by peer" error when peer TCP sends RST over a connection.
+// If peer chooses to send a FIN then we get the IsConnectionClosed() error else this.
+// Thrift fails with an error like:
+// [read tcp 10.0.0.7:33842->10.0.0.6:9090: read: connection reset by peer]
+//
+// Note: read()/recv() fails with this, i.e., only when the blobfuse process goes down after we successfully
+//
+//	      send the RPC request but before it could respond, the sender read/recv call fails with this error.
+//	      If the blobfuse process goes down before the send()/write() can send the RPC request, it fails
+//	      with IsBrokenPipe().
+//	      Another important thing to note is that IsBrokenPipe() can come when the target blobfuse process
+//		 might have restarted at some point in the past and may be running now, so it may make sense to
+//		 retry connection attempt on getting IsBrokenPipe(), whereas IsConnectionReset() means the process
+//		 just stopped/restarted, so trying connection attempt immediately may not help.
+func IsConnectionReset(err error) bool {
 	common.Assert(err != nil)
 
 	// RPC error, cannot be a connection reset error.
 	if GetRPCResponseError(err) != nil {
-		log.Debug("IsConnectionTerminated: is RPC error: %v", err)
+		log.Debug("IsConnectionReset: is RPC error: %v", err)
 		return false
 	}
 
+	te := thrift.NewTTransportExceptionFromError(err)
 	// Note: This doesn't work.
-	//te := thrift.NewTTransportExceptionFromError(err)
 	//return te.TypeId() == thrift.NOT_OPEN
+	log.Debug("IsConnectionReset: err: %v, err: %T, te.TypeId(): %d, Is syscall.ECONNRESET: %v",
+		err, err, te.TypeId(), errors.Is(err, syscall.ECONNRESET))
 
-	return errors.Is(err, syscall.EPIPE)
+	//
+	// In this case I've seen syscall.ECONNRESET check to work, but since it's not documented
+	// we do the string match also, just in case.
+	//
+	connectionResetByPeer := "onnection reset by peer"
```

Comment:
Is this intentional? I see this pattern in other places also where we are matching the error string.
```suggestion
	connectionResetByPeer := "connection reset by peer"
```